crates/pera-core/src/dwallet_mpc/mpc_manager.rs

```diff
@@ -0,0 +1,483 @@
+                .await?;
+        }
+        Ok(())
+    }
+
+    /// Update the encryption of decryption key share with the new shares.
+    /// This function is called when the network DKG protocol is done.
+    fn update_dwallet_mpc_network_key(
+        &self,
+        session_info: &SessionInfo,
+        public_output: MPCPublicOutput,
+        private_output: MPCPrivateOutput,
+    ) -> DwalletMPCResult<()> {
+        if let MPCRound::NetworkDkg(key_type) = session_info.mpc_round {
+            let epoch_store = self.epoch_store()?;
+            let network_keys = epoch_store
+                .dwallet_mpc_network_keys
+                .get()
+                .ok_or(DwalletMPCError::MissingDwalletMPCDecryptionKeyShares)?;
+
+            network_keys.add_key_version(
+                epoch_store.clone(),
+                key_type,
+                private_output.clone(),
+                public_output,
+            )?;
+        }
+        Ok(())
+    }
+
+    fn epoch_store(&self) -> DwalletMPCResult<Arc<AuthorityPerEpochStore>> {
+        self.epoch_store
+            .upgrade()
+            .ok_or(DwalletMPCError::EpochEnded(self.epoch_id))
+    }
+
+    /// Handles a message by forwarding it to the relevant MPC session.
+    /// If the session does not exist, punish the sender.
+    pub(crate) fn handle_message(
+        &mut self,
+        message: &[u8],
+        authority_name: AuthorityName,
+        session_id: ObjectID,
+    ) -> DwalletMPCResult<()> {
+        if self.malicious_actors.contains(&authority_name) {
+            return Ok(());
+        }
+        let session = match self.mpc_sessions.get_mut(&session_id) {
+            Some(session) => session,
+            None => {
+                warn!(
+                    "received a message for an MPC session ID: `{:?}` which does not exist",
+                    session_id
+                );
+                self.malicious_actors.insert(authority_name);
+                return Ok(());
+            }
+        };
+        match session.handle_message(&DWalletMPCMessage {
+            message: message.to_vec(),
+            authority: authority_name,
+        }) {
+            Err(DwalletMPCError::MaliciousParties(malicious_parties)) => {
+                self.flag_parties_as_malicious(&malicious_parties)?;
+                Ok(())
+            }
+            other => other,
+        }
+    }
+
+    /// Convert the indices of the malicious parties to their addresses and store them
+    /// in the malicious actors set.
+    /// New messages from these parties will be ignored.
+    /// todo(zeev): clarify if it's restarted on epoch change.
+    fn flag_parties_as_malicious(&mut self, malicious_parties: &[PartyID]) -> DwalletMPCResult<()> {
+        let malicious_parties_names = malicious_parties
+            .iter()
+            .map(|party_id| {
+                self.epoch_store()?
+                    .committee()
+                    .authority_by_index(*party_id as u32)
```

Comment:
here's a vulnerability for you: a party can act maliciously and blame another party for it. 

It should be `party_id - 1` since the indices are from `0..n` and party IDs are `1..=n`.

Please make sure this bug doesn't exist anywhere else in the code by having a dedicated function for converting to and through party IDs and indices.